### ISSUE TITLE
Consider adding in-place formatted code blocks

### ISSUE THREAD

**Issue by vlsi** [square/kotlinpoet#877]
The issue I have is positional arguments are sometimes hard to understand and maintain.
On the other hand, Kotlin has string templates, so why don't we use it for templating purposes?

For instance:

```kotlin
val hello = """Hello, "world", \n test"""
val j = FieldSpec.builder(Int::class.typeName, "j").build()

val codeBlock = codeBlock { "println(${hello.S} + ${j.N})" }

println(codeBlock.toString())
```

==>

```
println("Hello, \"world\", \\n test" + j)
```

One of my use cases:

```kotlin
codeBlock {
    """
    ${config.tokenTypeName.T} oldToken = ${token.N};
    ${config.tokenTypeName.T} nextToken = oldToken.next;
    if (nextToken == null) {
      nextToken = ${token_source.N}.${lexer.generateMethodName.N}(kind);
      ${jj_tokenCount.N}++;
      if (${tokenOutput.N} != null) {
          ${tokenOutput.N}.accept(nextToken);
      }
    }
    ${token.N} = nextToken;
    oldToken.next = null; // Stop nepotism
    return nextToken;

    """.trimIndent()
}
```

Here's my implementation (it is for JavaPoet's API, however, it should be more or less the same for KotlinPoet):

```kotlin
typealias StringTemplateMethod = StringTemplate.() -> String

class StringTemplate {
    val args = mutableListOf<Any?>()

    private fun arg(mask: String, v: Any?) = args.run {
        add(v)
        "$$size$mask"
    }

    val TypeName.T: String get() = arg("T", this)
    val FieldSpec.T: String get() = arg("T", this.type)
    val MethodSpec.T: String get() = arg("T", this.returnType)

    val FieldSpec.N: String get() = arg("N", this)
    val MethodSpec.N: String get() = arg("N", this)
    val TypeSpec.N: String get() = arg("N", this)

    val String.N: String get() = arg("N", this)

    val Any?.L: String get() = arg("L", this)
    val Any?.S: String get() = arg("S", this)
}

inline fun CodeBlock.Builder.add(template: StringTemplate.() -> String) = run {
    val t = StringTemplate()
    val format = t.template()
    add(format, *t.args.toTypedArray())!!
}

inline fun codeBlock(template: StringTemplate.() -> String) =
    CodeBlock.builder().add(template).build()!!
```

---

It might be convenient to have methods like `beginControlFlow(template: StringTemplateMethod)`

Basically every time there's `(String, vararg args)` method, it might be better written as `(StringTemplateMethod)`

For instance, the following sample from readme

```kotlin
.addStatement("if (ingredient is %T) taco %M ingredient", meat, minusAssign)
```

becomes

```kotlin
.addStatement { "if (ingredient is ${meat.T}) taco ${minusAssign.M} ingredient" }
```

---

See https://github.com/square/javapoet/issues/761#issuecomment-581057940
See https://github.com/tieskedh/KotlinPoetDSL/issues/55

**Comment by Egorand** [comment]
I don't think that proposed style is objectively better than positional arguments and I don't want the library to include both API styles. Thanks for the suggestion though!
